core/schema/module.go

```diff
@@ -721,6 +721,17 @@ func (s *moduleSchema) moduleWithSource(ctx context.Context, mod *core.Module, a
 		return nil, fmt.Errorf("failed to decode module source: %w", err)
 	}
 
+	engineVersion, err := src.Self.ModuleEngineVersion(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load module config: %w", err)
+	}
+	if err := engine.CheckVersionCompatibility(engineVersion, engine.MinimumModuleVersion); err != nil {
+		return nil, fmt.Errorf("module requires incompatible engine version: %w", err)
+	}
+	if err := engine.CheckVersionCompatibility(engine.Version, engineVersion); err != nil {
+		return nil, fmt.Errorf("module requires newer engine version: %w", err)
+	}
```

Comment:
How strict is this? Feels like it could be frustrating if someone isn't enabled to directly upgrade their Dagger version.

While we're pre-1.0 this is probably closer to the reality of our releases (patch releases introduce new features all the time that the module might depend on), but I'd expect post-1.0 when our versions are more meaningful that this would be way too strict.

I guess we could codify that ahead of time? If version is < 1.0, require strict `>=`, but if >= 1.0, allow wiggle room at the patch level?

Reply:
Ooh that's a good point. Hm, I think I agree - but we can encode that constraint when we get to v1.0 right? Since the released engine will be the one doing the checks.

Reply:
Yeah - as long as we remember. :smile: 

Maybe we should make a v1.0 milestone for things like that.

Reply:
Yeah I'm happy with that - aside from that, any outstanding issues?

Reply:
Opened https://github.com/dagger/dagger/issues/7865 (+ milestone) for the v1.0 change.

I think the main question is whether this is still too strict even for today, but I don't have a great read on that.